analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/remoteexecution/RemoteLatencyDataProvider.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2022 EngFlow Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.engflow.bazel.invocation.analyzer.dataproviders.remoteexecution;
+
+import com.engflow.bazel.invocation.analyzer.bazelprofile.BazelProfile;
+import com.engflow.bazel.invocation.analyzer.bazelprofile.BazelProfileConstants;
+import com.engflow.bazel.invocation.analyzer.core.DataProvider;
+import com.engflow.bazel.invocation.analyzer.core.DatumSupplier;
+import com.engflow.bazel.invocation.analyzer.core.DatumSupplierSpecification;
+import com.engflow.bazel.invocation.analyzer.core.InvalidProfileException;
+import com.engflow.bazel.invocation.analyzer.core.MissingInputException;
+import com.engflow.bazel.invocation.analyzer.core.NullDatumException;
+import com.google.common.annotations.VisibleForTesting;
+import java.time.Duration;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * A {@link DataProvider} that supplies an upper bound for the latency between the Bazel client and
+ * the remote execution/caching service. It does this estimation by taking half the minimum
+ * round-trip-time across all "remote action cache check" and "remote action execution" actions.
+ */
+public class RemoteLatencyDataProvider extends DataProvider {
+
+  @VisibleForTesting
+  public static final String EMPTY_REASON =
+      "The Bazel profile does not include any remote cache checks or executions, which are required"
+          + " for determining the latency. Try analyzing a profile for an invocation run against a"
+          + " remote execution service.";
+
+  @Override
+  public List<DatumSupplierSpecification<?>> getSuppliers() {
+    return List.of(
+        DatumSupplierSpecification.of(
+            RemoteLatency.class, DatumSupplier.memoized(this::getRemoteLatency)));
+  }
+
+  @VisibleForTesting
+  RemoteLatency getRemoteLatency()
+      throws InvalidProfileException, MissingInputException, NullDatumException {
+    BazelProfile bazelProfile = getDataManager().getDatum(BazelProfile.class);
+    Optional<Duration> duration =
+        bazelProfile
+            .getThreads()
+            .flatMap((thread) -> thread.getCompleteEvents().stream())
+            .filter(
+                (event) ->
+                    BazelProfileConstants.CAT_REMOTE_ACTION_EXECUTION.equals(event.category)
```

Comment:
I would honestly exclude this, remote execution is such an involved step that even the fastest of the remote exec actions will most likely have a significant amount of time in addition to latency. Not to say that roughly estimating latency from remote exec actions is impossible, but looking at the minimum simply would not suffice. Can we just rely on action cache checks for now?

Reply:
Will there always be action cache checks? If they will always be faster, the remote exec actions are just a fall-back, since it takes them minimum. These two actions were specified in the requirements in EngFlow/engflow#9404. @saraadams WDYT?

Reply:
I'd rather you ignored this comment and left a comment in the code explaining the issue than leaving the PR open for so long :)

Reply:
Comment added! This was a rather unusual situation due to Sara being out, and me being partially out.

Reply:
> Will there always be action cache checks? If they will always be faster, the remote exec actions are just a fall-back, since it takes them minimum. These two actions were specified in the requirements in [EngFlow/engflow#9404](https://github.com/EngFlow/engflow/issues/9404). @saraadams WDYT?

The current solution you implemented (keep remote execution actions as a fall-back) is what I would have recommended.
(Note that some actions could be executed super fast, and latency may vary. So even if there are remote cache checks, a remote execution action might be faster. That's not the pattern we are looking for here, but I still thought it worth noting.)